src/coreclr/jit/hwintrinsiccodegenarm64.cpp

```diff
@@ -364,7 +364,44 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
         {
             assert(!hasImmediateOperand);
 
-            switch (intrin.numOperands)
+            size_t numOperands = intrin.numOperands;
+
+            // This handles optimizations for instructions that have
+            // an implicit 'zero' vector of what would be the second operand.
+            if ((numOperands == 2) && ((intrin.op2->IsVectorZero() && intrin.op2->isContained()) ||
+                                       (intrin.op1->IsVectorZero() && intrin.op1->isContained() &&
+                                        HWIntrinsicInfo::IsCommutative(intrin.id))))
+            {
+                assert(HWIntrinsicInfo::SupportsContainment(intrin.id));
+
+                if (intrin.op1->IsVectorZero() && intrin.op1->isContained() &&
+                    HWIntrinsicInfo::IsCommutative(intrin.id))
+                {
+                    // The intrinsic is commutative, swap the registers.
+                    assert(op1Reg == REG_NA);
+                    op1Reg = op2Reg;
+                    op2Reg = REG_NA;
+                }
+
+                switch (ins)
+                {
+                    case INS_cmeq:
+                    case INS_cmge:
+                    case INS_cmgt:
+                    case INS_fcmeq:
+                    case INS_fcmge:
+                    case INS_fcmgt:
+                    {
+                        numOperands -= 1;
```

Comment:
Would it be clearer to just set this to `1`?

And would it be better to assert its one of these instructions rather than doing it in a switch? I expect the perf here doesn't matter much, but it would be "cheaper" that way.

Reply:
Yea, probably :)